File: libs/go/radar.go
package stonecharts

import (
	"fmt"
	"math"
	"strings"
	"unicode/utf8"
)

func renderRadarSVG(spec *ChartSpec) string {
	W, H := int(spec.Width), int(spec.Height)
	theme := spec.theme
	if theme == nil {
		t := lightTheme()
		theme = &t
	}
	palette := theme.Palette
	cid := esc(spec.ID)
	_ = cid

	a11yAttr := ""
	a11yDesc := ""
	if spec.a11yOn() {
		sum := esc(a11ySummary(spec, "Radar"))
		a11yAttr = fmt.Sprintf(` role="img" aria-label="%s"`, sum)
		a11yDesc = fmt.Sprintf("<desc>%s</desc>", sum)
	}

	mTop := 20.0
	if spec.Title != "" {
		mTop += 26
	}
	if spec.Subtitle != "" {
		mTop += 18
	}
	mLeft := 22.0
	mRight := 22.0
	mBottom := 28.0
	if spec.legendOn() {
		mBottom += 18
	}
	if spec.Layout != nil && spec.Layout.Margin != nil {
		m := spec.Layout.Margin
		if m.Top != nil {
			mTop = *m.Top
		}
		if m.Left != nil {
			mLeft = *m.Left
		}
		if m.Right != nil {
			mRight = *m.Right
		}
		if m.Bottom != nil {
			mBottom = *m.Bottom
		}
	}

	plotX, plotY := mLeft, mTop
	plotW, plotH := float64(W)-mLeft-mRight, float64(H)-mTop-mBottom

	cats := spec.XAxis.Categories
	nAxes := len(cats)
	if nAxes < 3 {
		nAxes = 3
	}

	yMin := 0.0
	if spec.YAxis.Min != nil {
		yMin = *spec.YAxis.Min
	}
	yMax := 0.0
	hasYMax := false
	if spec.YAxis.Max != nil {
		yMax = *spec.YAxis.Max
		hasYMax = true
	}
	if !hasYMax {
		for _, s := range spec.Series {
			for _, v := range s.Data {
				if v > yMax {
					yMax = v
				}
			}
		}
	}
	if yMax <= yMin {
		yMax = yMin + 100
	}

	labelMargin := 40.0
	cx := plotX + plotW/2
	cy := plotY + plotH/2
	rMax := plotW / 2
	if plotH/2 < rMax {
		rMax = plotH / 2
	}
	rMax -= labelMargin

	nRings := 5

	axisAngle := func(i int) float64 {
		return -math.Pi/2 + float64(i)*2*math.Pi/float64(nAxes)
	}
	pointAt := func(angle, r float64) (float64, float64) {
		return cx + r*math.Cos(angle), cy + r*math.Sin(angle)
	}

	var p strings.Builder

	font := `font-family="Segoe UI, Helvetica, Arial, sans-serif"`
	if spec.Responsive {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %d %d" preserveAspectRatio="xMidYMid meet" width="100%%" %s>`,
			a11yAttr, W, H, font))
	} else {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d" %s>`,
			a11yAttr, W, H, W, H, font))
	}

	if a11yDesc != "" {
		p.WriteString(a11yDesc)
	}

	if theme.Background != "" {
		p.WriteString(fmt.Sprintf(
			`<rect class="sc-bg" x="0" y="0" width="%d" height="%d" fill="%s"/>`,
			W, H, theme.Background))
	}

	ty := 26
	if spec.Title != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-title" x="%s" y="%d" text-anchor="middle" font-size="17" font-weight="600" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.TitleColor, esc(spec.Title)))
		ty += 20
	}
	if spec.Subtitle != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-subtitle" x="%s" y="%d" text-anchor="middle" font-size="12" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.SubtitleColor, esc(spec.Subtitle)))
	}

	p.WriteString(fmt.Sprintf(
		`<line class="sc-crosshair" x1="0" y1="%s" x2="0" y2="%s" stroke="%s" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>`,
		f1(plotY), f1(plotY+plotH), theme.CrosshairColor))

	// Grid rings (spiderweb)
	for level := 0; level < nRings; level++ {
		frac := float64(level+1) / float64(nRings)
		r := rMax * frac
		var pts strings.Builder
		for i := 0; i < nAxes; i++ {
			if i > 0 {
				pts.WriteString(" ")
			}
			px, py := pointAt(axisAngle(i), r)
			pts.WriteString(fmt.Sprintf("%s,%s", f1(px), f1(py)))
		}
		p.WriteString(fmt.Sprintf(
			`<polygon class="sc-radar-ring" data-level="%d" points="%s" fill="none" stroke="%s" stroke-width="1"/>`,
			level, pts.String(), theme.GridColor))
	}

	// Radial axis lines
	for i := 0; i < nAxes; i++ {
		ex, ey := pointAt(axisAngle(i), rMax)
		p.WriteString(fmt.Sprintf(
			`<line class="sc-radar-axis" data-index="%d" x1="%s" y1="%s" x2="%s" y2="%s" stroke="%s" stroke-width="1"/>`,
			i, f1(cx), f1(cy), f1(ex), f1(ey), theme.GridColor))
	}

	// Category labels
	for i := 0; i < nAxes; i++ {
		angle := axisAngle(i)
		lx, ly := pointAt(angle, rMax+12)
		cosA := math.Cos(angle)
		anchor := "middle"
		if cosA > 0.01 {
			anchor = "start"
		} else if cosA < -0.01 {
			anchor = "end"
		}
		label := fmt.Sprintf("%d", i)
		if i < len(cats) {
			label = cats[i]
		}
		p.WriteString(fmt.Sprintf(
			`<text class="sc-radar-label" data-index="%d" x="%s" y="%s" text-anchor="%s" dominant-baseline="central" font-size="11" fill="%s">%s</text>`,
			i, f1(lx), f1(ly), anchor, theme.AxisLabelColor, esc(label)))
	}

	// Tick labels along first axis
	for level := 0; level < nRings; level++ {
		frac := float64(level+1) / float64(nRings)
		tickVal := yMin + frac*(yMax-yMin)
		tickR := rMax * frac
		tx, ty2 := pointAt(axisAngle(0), tickR)
		p.WriteString(fmt.Sprintf(
			`<text class="sc-radar-tick" data-value="%s" x="%s" y="%s" font-size="9" fill="%s">%s</text>`,
			esc(fmtNum(tickVal)), f1(tx+4), f1(ty2-2), theme.AxisLabelColor, esc(fmtNum(tickVal))))
	}

	// Series polygons and vertices
	for si, s := range spec.Series {
		color := palette[si%len(palette)]
		grad, solidHex := s.colorSpec()
		if grad != nil {
			if len(grad.Stops) > 0 {
				color = grad.Stops[0].Color
			}
		} else if solidHex != "" {
			color = solidHex
		}
		color = esc(color)

		fillOpacity := s.FillOpacity

		type vertex struct{ x, y float64 }
		vertices := make([]vertex, nAxes)
		for j := 0; j < nAxes; j++ {
			v := 0.0
			if j < len(s.Data) {
				v = s.Data[j]
			}
			frac := (v - yMin) / (yMax - yMin)
			if frac < 0 {
				frac = 0
			}
			if frac > 1 {
				frac = 1
			}
			r := rMax * frac
			vx, vy := pointAt(axisAngle(j), r)
			vertices[j] = vertex{vx, vy}
		}

		var pathD strings.Builder
		for k, vt := range vertices {
			if k == 0 {
				pathD.WriteString(fmt.Sprintf("M %s %s", f1(vt.x), f1(vt.y)))
			} else {
				pathD.WriteString(fmt.Sprintf(" L %s %s", f1(vt.x), f1(vt.y)))
			}
		}
		pathD.WriteString(" Z")

		fillAttr := `fill="none"`
		opacityAttr := ""
		if fillOpacity > 0 {
			fillAttr = fmt.Sprintf(`fill="%s"`, color)
			opacityAttr = fmt.Sprintf(` fill-opacity="%g"`, fillOpacity)
		}

		p.WriteString(fmt.Sprintf(
			`<path class="sc-radar-poly sc-point" data-series="%d" data-series-name="%s" data-color="%s" d="%s" %s%s stroke="%s" stroke-width="2"/>`,
			si, esc(s.Name), color, pathD.String(), fillAttr, opacityAttr, color))

		for j, vt := range vertices {
			v := 0.0
			if j < len(s.Data) {
				v = s.Data[j]
			}
			p.WriteString(fmt.Sprintf(
				`<circle class="sc-radar-dot sc-point" data-series="%d" data-index="%d" data-y="%s" cx="%s" cy="%s" r="4" fill="%s"/>`,
				si, j, esc(fmtNum(v)), f1(vt.x), f1(vt.y), color))
		}
	}

	// Legend
	if spec.legendOn() && len(spec.Series) > 0 {
		gap := 22.0
		est := make([]float64, len(spec.Series))
		for i, s := range spec.Series {
			est[i] = float64(utf8.RuneCountInString(s.Name)*7 + 26)
		}
		totalW := 0.0
		for _, e := range est {
			totalW += e
		}
		totalW += gap * float64(len(est)-1)
		lx := plotX + (plotW-totalW)/2
		ly := float64(H - 10)
		p.WriteString(`<g class="sc-legend">`)
		for si, s := range spec.Series {
			color := palette[si%len(palette)]
			grad, solidHex := s.colorSpec()
			if grad != nil {
				if len(grad.Stops) > 0 {
					color = grad.Stops[0].Color
				}
			} else if solidHex != "" {
				color = solidHex
			}
			color = esc(color)
			p.WriteString(fmt.Sprintf(`<g class="sc-legend-item" data-series="%d">`, si))
			p.WriteString(fmt.Sprintf(
				`<rect x="%s" y="%s" width="14" height="4" rx="2" fill="%s"/>`,
				f1(lx), f1(ly-9), color))
			p.WriteString(fmt.Sprintf(
				`<text x="%s" y="%s" font-size="12" fill="%s">%s</text>`,
				f1(lx+20), f1(ly-2), theme.LegendTextColor, esc(s.Name)))
			p.WriteString("</g>")
			lx += est[si] + gap
		}
		p.WriteString("</g>")
	}

	p.WriteString("</svg>")
	return p.String()
}

func radarDataTable(spec *ChartSpec) string {
	var b strings.Builder
	b.WriteString(`<table class="sc-visually-hidden">`)
	if spec.Title != "" {
		b.WriteString("<caption>" + esc(spec.Title) + "</caption>")
	}
	cats := spec.XAxis.Categories
	n := 0
	for _, s := range spec.Series {
		if len(s.Data) > n {
			n = len(s.Data)
		}
	}
	b.WriteString("<thead><tr><td></td>")
	for i := 0; i < n; i++ {
		label := fmt.Sprintf("%d", i)
		if i < len(cats) {
			label = cats[i]
		}
		b.WriteString(`<th scope="col">` + esc(label) + `</th>`)
	}
	b.WriteString("</tr></thead><tbody>")
	for _, s := range spec.Series {
		b.WriteString(`<tr><th scope="row">` + esc(s.Name) + `</th>`)
		for i := 0; i < n; i++ {
			if i < len(s.Data) {
				b.WriteString("<td>" + esc(fmtNum(s.Data[i])) + "</td>")
			} else {
				b.WriteString("<td></td>")
			}
		}
		b.WriteString("</tr>")
	}
	b.WriteString("</tbody></table>")
	return b.String()
}

File: libs/python/stonecharts/charts/radar.py
"""Radar / spider chart renderer: ChartSpec -> SVG string.

Multi-dimensional categorical data on radial axes forming polygonal overlays.
Non-Cartesian (Family B polar sibling) — own SVG shell, no axes.
See charts/radar/design.md for the full geometry contract.
"""

from __future__ import annotations

import math

from ..spec import ChartSpec, Gradient
from ..util import esc, fmt_num
from ._cartesian import a11y_summary


def render_svg(spec: ChartSpec) -> str:
    W, H = spec.width, spec.height
    theme = spec.theme
    palette = theme.palette
    _cid = esc(spec.id)

    a11y_attr = ""
    a11y_desc = ""
    if spec.a11y:
        _sum = esc(a11y_summary(spec, "Radar"))
        a11y_attr = f' role="img" aria-label="{_sum}"'
        a11y_desc = f"<desc>{_sum}</desc>"

    m_top: float = 20
    if spec.title:
        m_top += 26
    if spec.subtitle:
        m_top += 18
    m_left: float = 22
    m_right: float = 22
    m_bottom: float = 28 + (18 if spec.legend else 0)
    if spec.layout and spec.layout.margin:
        m = spec.layout.margin
        if m.top is not None:
            m_top = m.top
        if m.left is not None:
            m_left = m.left
        if m.right is not None:
            m_right = m.right
        if m.bottom is not None:
            m_bottom = m.bottom

    plot_x, plot_y = m_left, m_top
    plot_w, plot_h = W - m_left - m_right, H - m_top - m_bottom

    cats = spec.x_axis.categories or []
    n_axes = len(cats)
    if n_axes < 3:
        n_axes = 3

    y_min = spec.y_axis.min if spec.y_axis.min is not None else 0.0
    y_max_spec = spec.y_axis.max
    if y_max_spec is not None:
        y_max = y_max_spec
    else:
        y_max = 0.0
        for s in spec.series:
            for v in s.data:
                if v > y_max:
                    y_max = v
    if y_max <= y_min:
        y_max = y_min + 100

    label_margin = 40.0
    cx = plot_x + plot_w / 2
    cy = plot_y + plot_h / 2
    r_max = min(plot_w, plot_h) / 2 - label_margin

    n_rings = 5

    p: list[str] = []

    _font = 'font-family="Segoe UI, Helvetica, Arial, sans-serif"'
    if spec.responsive:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'viewBox="0 0 {W} {H}" preserveAspectRatio="xMidYMid meet" '
            f'width="100%" {_font}>'
        )
    else:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'width="{W}" height="{H}" viewBox="0 0 {W} {H}" {_font}>'
        )

    if a11y_desc:
        p.append(a11y_desc)

    if theme.background:
        p.append(f'<rect class="sc-bg" x="0" y="0" width="{W}" height="{H}" fill="{theme.background}"/>')

    ty = 26
    if spec.title:
        p.append(
            f'<text class="sc-title" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="17" font-weight="600" fill="{theme.title_color}">{esc(spec.title)}</text>'
        )
        ty += 20
    if spec.subtitle:
        p.append(
            f'<text class="sc-subtitle" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="12" fill="{theme.subtitle_color}">{esc(spec.subtitle)}</text>'
        )

    p.append(
        f'<line class="sc-crosshair" x1="0" y1="{plot_y:.1f}" x2="0" y2="{plot_y + plot_h:.1f}" '
        f'stroke="{theme.crosshair_color}" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>'
    )

    def axis_angle(i: int) -> float:
        return -math.pi / 2 + i * 2 * math.pi / n_axes

    def point_at(angle: float, r: float) -> tuple[float, float]:
        return cx + r * math.cos(angle), cy + r * math.sin(angle)

    for level in range(n_rings):
        frac = (level + 1) / n_rings
        r = r_max * frac
        pts = " ".join(
            f"{point_at(axis_angle(i), r)[0]:.1f},{point_at(axis_angle(i), r)[1]:.1f}" for i in range(n_axes)
        )
        p.append(
            f'<polygon class="sc-radar-ring" data-level="{level}" '
            f'points="{pts}" fill="none" stroke="{theme.grid_color}" stroke-width="1"/>'
        )

    for i in range(n_axes):
        ex, ey = point_at(axis_angle(i), r_max)
        p.append(
            f'<line class="sc-radar-axis" data-index="{i}" '
            f'x1="{cx:.1f}" y1="{cy:.1f}" x2="{ex:.1f}" y2="{ey:.1f}" '
            f'stroke="{theme.grid_color}" stroke-width="1"/>'
        )

    for i in range(n_axes):
        angle = axis_angle(i)
        lx, ly = point_at(angle, r_max + 12)
        cos_a = math.cos(angle)
        if cos_a > 0.01:
            anchor = "start"
        elif cos_a < -0.01:
            anchor = "end"
        else:
            anchor = "middle"
        label = cats[i] if i < len(cats) else str(i)
        p.append(
            f'<text class="sc-radar-label" data-index="{i}" '
            f'x="{lx:.1f}" y="{ly:.1f}" text-anchor="{anchor}" '
            f'dominant-baseline="central" font-size="11" fill="{theme.axis_label_color}">'
            f"{esc(label)}</text>"
        )

    for level in range(n_rings):
        frac = (level + 1) / n_rings
        tick_val = y_min + frac * (y_max - y_min)
        tick_r = r_max * frac
        tx, ty2 = point_at(axis_angle(0), tick_r)
        p.append(
            f'<text class="sc-radar-tick" data-value="{esc(fmt_num(tick_val))}" '
            f'x="{tx + 4:.1f}" y="{ty2 - 2:.1f}" font-size="9" '
            f'fill="{theme.axis_label_color}">{esc(fmt_num(tick_val))}</text>'
        )

    for si, s in enumerate(spec.series):
        color = palette[si % len(palette)]
        if isinstance(s.color, Gradient):
            color = s.color.stops[0].color if s.color.stops else color
        elif s.color:
            color = s.color
        color = esc(color)

        fill_opacity = s.fill_opacity

        vertices: list[tuple[float, float]] = []
        for j in range(n_axes):
            v = s.data[j] if j < len(s.data) else 0.0
            frac = (v - y_min) / (y_max - y_min)
            frac = max(0.0, min(1.0, frac))
            r = r_max * frac
            vx, vy = point_at(axis_angle(j), r)
            vertices.append((vx, vy))

        path_d = " ".join(f"{'M' if k == 0 else 'L'} {vx:.1f} {vy:.1f}" for k, (vx, vy) in enumerate(vertices))
        path_d += " Z"

        fill_attr = f'fill="{color}"' if fill_opacity > 0 else 'fill="none"'
        opacity_attr = f' fill-opacity="{fill_opacity}"' if fill_opacity > 0 else ""

        p.append(
            f'<path class="sc-radar-poly sc-point" data-series="{si}" '
            f'data-series-name="{esc(s.name)}" data-color="{color}" '
            f'd="{path_d}" {fill_attr}{opacity_attr} '
            f'stroke="{color}" stroke-width="2"/>'
        )

        for j, (vx, vy) in enumerate(vertices):
            v = s.data[j] if j < len(s.data) else 0.0
            p.append(
                f'<circle class="sc-radar-dot sc-point" data-series="{si}" '
                f'data-index="{j}" data-y="{esc(fmt_num(v))}" '
                f'cx="{vx:.1f}" cy="{vy:.1f}" r="4" fill="{color}"/>'
            )

    if spec.legend and spec.series:
        gap = 22
        est = [len(s.name) * 7 + 26 for s in spec.series]
        total_w = sum(est) + gap * (len(est) - 1) if est else 0
        lx = plot_x + (plot_w - total_w) / 2
        ly = H - 10
        p.append('<g class="sc-legend">')
        for si, s in enumerate(spec.series):
            color = palette[si % len(palette)]
            if isinstance(s.color, Gradient):
                color = s.color.stops[0].color if s.color.stops else color
            elif s.color:
                color = s.color
            color = esc(color)
            p.append(f'<g class="sc-legend-item" data-series="{si}">')
            p.append(f'<rect x="{lx:.1f}" y="{ly - 9:.1f}" width="14" height="4" rx="2" fill="{color}"/>')
            p.append(
                f'<text x="{lx + 20:.1f}" y="{ly - 2:.1f}" font-size="12" '
                f'fill="{theme.legend_text_color}">{esc(s.name)}</text>'
            )
            p.append("</g>")
            lx += est[si] + gap
        p.append("</g>")

    p.append("</svg>")
    return "".join(p)
